### DISCUSSION TITLE
Format for TRS-80 35 Track SS/SD 5.25 floppies

### DISCUSSION THREAD

**Discussion by DrRGee** [keirf/greaseweazle#306] [General]
I am struggling trying to read old TRS-80 floppies on a Shugart SA400.. Using ibm.fm all appears to work fine works fine EXCEPT for T17, which is the directory. I believe that the sector DAM=0xFB everywhere else, but for the directory track it is 0xFA. Looking at ibm.py L126, it looks like DAM is set as 0xFB - is there any hope to be able to specify the DAM or can you suggest another possible solution?

**Comment by keirf** [comment]
I think any directory sector anywhere on disk is DAM FA? What format are you converting to? Eg DAM byte info would be lost when converting to IMG. Otherwise simplest thing to do is accept FA as a DAM value in ibm.py.

**Reply by DrRGee** [reply]
Thanks for responding. Yes, my understanding of the 35T SS/SD TRS-80 format is that any sector with a DAM=0xFA is a directory sector, although, T17 is where they normally appear. Are you saying to rewrite a fork to allow both FA or FB as a DAM?

**Reply by keirf** [reply]
I wonder how important the FA mark is to preserve? FlashFloppy supports TRS-80 by all accounts, but certainly never emits an FA mark when used with IMG files.

Anyway yes I'm basically proposing that `fm_decode_raw` becomes a bit more permissive. Let me make a quick change.

**Reply by keirf** [reply]
Any idea if that mark is ever used in MFM tracks?

**Reply by keirf** [reply]
Take a look at and try out branch `trs`. You can also see a Windows prebuilt here: https://github.com/keirf/greaseweazle/actions/runs/4844962541

(Look at bottom under artifacts. You need to be logged into GitHub to see the link.)

**Reply by jdmba** [reply]
@keirf - DAM's are just the tip of the iceberg.  The TRS-80 has many copy types of copy protection which was used over the years.  I don't think the TRS-80 can be specifically written for; but rather needs a lot of flexibility.   Even without copy protection, mixed FM and MFM is a common format.  I can certainly provide DMK images of the wide variety of TRS-80 formats out there, but am not experienced enough to answer technical questions.

**Reply by jdmba** [reply]
@keirf - There is an excellent write up of the disk formats for the STANDARD TRS-80 disks, including the DAM's, here https://www.tim-mann.org/trs80/dskspec.html

**Comment by DrRGee** [comment]
Will Do!
________________________________

**Comment by DrRGee** [comment]
OK, here is what I did last night using 1.10

gw.exe read --diskdefs "C:\TRS80\greaseweazle-1.10\diskdefs.cfg" --format=TRSSSD.80 --retries=5 --revs=5 --drive=1 --tracks=c=0-34 "C:\TRS80\greaseweazle-1.10\mydisk5ti9.img"

and that yielded


[cid:0865786d-506a-45f9-ad18-ba2205a05752]

With all of the sectors on T17 missing.

Then, I installed the zip. Moved the guitools into that directory and used:
gw.exe read --diskdefs "C:\TRS80\greaseweazle-1.10\diskdefs.cfg" --format=TRSSSD.80 --retries=5 --revs=5 --drive=1 --tracks=c=0-34 "C:\TRS80\greaseweazle-1.10\mydisk5ti9.img"

and that yielded this!
[cid:acd0ee95-be3c-40dd-9e57-5d913893237e]

TRSSSD.80 is an entry for the cfg I was working on:
and contains:
disk TRSSSD.80
    cyls = 35
    heads = 1
    tracks * ibm.fm
        secs = 10
      id=0
        bps = 256
        gap1=20
      gap2=20
      #gap3 = 84
      gap3=30
        rate = 125
    end
end

The gaps are not right. I was using https://archive.org/details/insup/page/n11/mode/2up SuperUtility Plus documentation, see page 8- but it got late.

I want to look at the .img and then your code but this is VERY encouraging!

Thank you.
[https://archive.org/services/img/insup]<https://archive.org/details/insup/page/n11/mode/2up>
Inside Super Utility Plus : PowerSoft : Free Download, Borrow, and Streaming : Internet Archive<https://archive.org/details/insup/page/n11/mode/2up>
Inside Super Utility Plus
archive.org

________________________________

**Reply by keirf** [reply]
GitHub strips attachments from emails so disk images are missing. I wouldn't trust documentation for the gaps values. Especially gap3. Other gaps are more standardised. You will need to work out gap3 by trial and error right now, or send me a flux dump and I can work it out.

**Comment by DrRGee** [comment]
Yep, I have some work to do, but it is good work...quite a giggle seeing a Z-80 program that I wrote some 40 years ago 🙂

I have a 40T drive and I should be able to look at some newdos and MIII disks.

very​ cool!

Thanks so much.


________________________________

**Reply by TimothyPMann** [reply]
Model I single density (not double) wants an FA DAM on the directory track, which is normally 17 but can be another track. TRSDOS won't be able to find the directory if it's written with another DAM. So if you're creating Model I TRSDOS disks you'll need to be able to write it.

Model 3/4/4P can't write that DAM or distinguish it on read from FB. They generally use F8 on the directory track. Double density disks written on Model I with a double density adapter also use F8. Model I LDOS will tolerate F8 DAM in single density, but TRSDOS won't unless patched

**Comment by Verault** [comment]
@DrRGee Just curious how this panned out for you. The TRSSSD.80 config file you created seems like a good idea and I was wondering if you figured out the missing variables (gaps) and have a working config as of now. I have bought a kryoflux 10 years ago but a fellow enthusiast asked me to build a Greaseweazle too (as he has one) so I can help him out with Disk creation on the GW. Its built and time to start testing so I was wondering what you found out.

**Reply by DrRGee** [reply]
It has worked out rather well for me but I had some specific things in mind; verify that my 40+ year old drives work (and clean and adjust timing). See if I could read *any* of the original diskettes that I have. The answer is yes on both counts. 

The newest version that incorporated the sector mark for the original TRSDOS format is a big plus (v1.11). I can read almost all of my old M1 TRSDOS disks that I have tried (maybe 25-30) and many that I have tried are error free AFTER I clean them and the drive head. (read to img and dsk formats). Of, course, what I would really like is to be able to specify the sector id in the config file. 

I can not read any of my NEWDOS diskettes at all - I don't know how I formatted them and they probably have unsupported sector marks. 

I wish the software was written in C/C++, but it is, IMO, a good product and am happy with the purchase.

Of course, I am putting off resurrecting the actual machine and when I do, I should be able to read everything as was initially intended :)

As for the gaps; I messed around with them, but I never systematically evaluated how to optimize them - mainly because I don't know how (although I thought those disk maps were pretty cool). Again, it would seem like I could easily get the actual data once I have the machine working.

**Comment by DrRGee** [comment]
To update/conclude my original message.

Since embarking on the archive project and first posting here, I have now used GW (V1.5/6 firmware) to read (and sometimes write) ~100, TRS-80 MI SSSD 35-Track disks, not copy-protected but that were some 40+ years old.

Mostly these were TRS-DOS 2.3 disks. I resuscitated a TRS-80 M1 with EI and 5.25 drives and was able to test out a great many issues. 

Of course, not every disk read perfectly and a good deal of cleaning was required – you can get plenty of info on that elsewhere.

I archived using the GW .dsk format. This was chosen because TRSTools, which offers many advantages, can use those .dsk files. 

I wrote many disks (using NOS floppies) using GW and the .dsk file that was read from the old disks. These will write just fine **BUT**, you can’t read them straight away on a TRS80 M1 because of the DAM (there is plenty of info on that). Because I had a working system, I could use Super Utility to easily write the correct DAM on T17 and, after that, they read just fine on the TRS80 M1.

I used several different drives, like the SA400 and TM100-1/2.
This is the script (I basically posted this before) that I used for the bulk of the copies. I will not quibble about the GAP sizes but instead just say that this worked for me, over and over again.

TRS-DOS 2.3 tested
TRS-80 Directory Address Mark  can be read but NOT written

disk TRS80SSSDa.35T
    cyls = 35
    heads = 1
    tracks * ibm.fm
        secs = 10
        id=0
        bps = 256
        gap1=18
        gap2=18
        gap3=20
        rate = 125
        interleave = 2
    end
end

GW is an economical and excellent tool.

My wish list:

Flexible DAM writing.

.dmk support (yes, I read the discussion posts) for MIII

.dsk for MIII (I still can’t do this and get a .jv3 format that is readable by TRSTools)

Thanks so much for making this tool available!

**Reply by qbarnes** [reply]
I've got `gw2dmk` (https://github.com/qbarnes/gw2dmk) working with basic DMK read and write functionality these days.  It re-implements @TimothyPMann 's `cw2dmk` (https://github.com/qbarnes/cw2dmk) for the Greaseweazle.  There are no man pages for the binaries yet, but their usage messages are pretty thorough, and for the most part, you can use the man pages from `cw2dmk`/`dmk2cw`.

I have binary releases available for ARM and x86_64 for Linux as well as MS Windows 32-/64-bit.  The binaries should be able to read and write pretty much any TRS-80 format out there to and from DMK format.  (Anything that `cw2dmk` could handle, `gw2dmk` should as well.)  If you have the chance to try it out and find any bugs or have questions, please let me know.

**Reply by DrRGee** [reply]
I certainly would like to try it! Is there a link for the MS Windows 32-/64-bit binaries?

**Reply by qbarnes** [reply]
I should have included the link. Here you go: https://github.com/qbarnes/gw2dmk/releases

**Reply by jdmba** [reply]
@DrRGee - If you have any images which you think might have software that can be added to the TRS-80 archive, please consider sharing!
